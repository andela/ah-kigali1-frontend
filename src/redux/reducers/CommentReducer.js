/* eslint-disable no-case-declarations */
import {
  SET_SUCCESS_MESSAGE,
  SET_ERROR_MESSAGE,
  COMMENTS_INPUT,
  COMMENTS_INPUT_EDIT,
  FETCHING_COMMENTS,
  UPDATE_COMMENT_BODY,
  SET_LOADING_COMMENTS,
  DELETE_COMMENT,
  UPDATE_COMMENT,
  CREATE_NEW_COMMENT
} from "../actionTypes";

const initialState = {
  body: "",
  bodyEdit: "",
  success: "",
  error: "",
  comments: {},
  isLoading: true
};

const commentReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_LOADING_COMMENTS:
      return {
        ...state,
        isLoading: true
      };
    case FETCHING_COMMENTS:
      return {
        ...state,
        isLoading: false,
        comments:
          action.page === 1
            ? action.comments
            : { ...state.comments, ...action.comments }
      };
    case COMMENTS_INPUT_EDIT:
      return {
        ...state,
        [action.payload.field]: action.payload.value
      };
    case UPDATE_COMMENT_BODY:
      return { ...state, bodyEdit: action.payload.value };
    case SET_SUCCESS_MESSAGE:
      return { ...state, success: action.payload };
    case SET_ERROR_MESSAGE:
      return { ...state, isLoading: false, error: action.payload };
    case COMMENTS_INPUT:
      return { ...state, [action.payload.field]: action.payload.value };
    case DELETE_COMMENT:
      delete state.comments[action.payload];
      return {
        ...state
      };
    case UPDATE_COMMENT:
      const { commentId, body } = action.payload;
      state.comments[commentId].body = body;
      return {
        ...state
      };
    case CREATE_NEW_COMMENT:
      return {
        ...state,
        comments: { [action.payload.id]: action.payload, ...state.comments }
      };
    default:
      return {
        ...state
      };
  }
};

export default commentReducer;
