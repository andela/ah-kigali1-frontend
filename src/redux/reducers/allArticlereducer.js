const initialState = {
  allArticles: []
};

export default (state = initialState, action) => {
  switch (action.type) {
    case "ALL_ARTICLES":
      return {
        ...state,
        allArticles: action.payload
      };
    case "FETCHING_ALL_ARTICLES":
      return {
        ...state,
        isFetching: true
      };
    case "ALL_ARTICLES_ERROR":
      return {
        ...state,
        ...action.payload
      };
    default:
      return {
        ...state
      };
  }
};
