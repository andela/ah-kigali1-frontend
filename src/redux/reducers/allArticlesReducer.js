import {
  FETCHING_ALL_ARTICLES,
  ALL_ARTICLES,
  ALL_ARTICLES_ERROR
} from "../actionTypes";

export const initialState = {
  allArticles: []
};

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case ALL_ARTICLES:
      return {
        ...state,
        allArticles: payload
      };
    case FETCHING_ALL_ARTICLES:
      return {
        ...state,
        isFetching: true
      };
    case ALL_ARTICLES_ERROR:
      return {
        ...state,
        payload
      };
    default:
      return state;
  }
};
