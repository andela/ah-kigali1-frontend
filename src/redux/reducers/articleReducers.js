import {
  ARTICLE_FETCHED,
  ARTICLE_ERROR,
  FETCHING_ASIDE_ARTICLES,
  FETCHING_ARTICLE,
  DELETE_ARTICLE
} from "../actionTypes";

const initialState = { isFetching: true, asideArticles: { articles: [] } };

export const articleReducer = (state = initialState, action) => {
  switch (action.type) {
    case ARTICLE_FETCHED:
      return {
        ...state,
        ...action.payload,
        isFetching: false
      };
    case FETCHING_ASIDE_ARTICLES:
      return {
        ...state,
        asideArticles: action.payload
      };
    case ARTICLE_ERROR:
      return {
        ...state,
        ...action.payload,
        isFetching: false
      };
    case FETCHING_ARTICLE:
      return {
        ...state,
        isFetching: true
      };
    case DELETE_ARTICLE:
      return {
        ...state,
        response: action.payload
      };
    default:
      return {
        ...state
      };
  }
};
