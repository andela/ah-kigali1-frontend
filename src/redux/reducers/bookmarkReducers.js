import {
  BOOKMARK_ARTICLE,
  REQUEST_SUBMITTED,
  REQUEST_FAILED
} from "../actionTypes";

export const initialState = {
  isSubmitting: false,
  message: "",
  bookmarkedArticles: {}
};

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case BOOKMARK_ARTICLE:
      return {
        ...state,
        bookmarkedArticles: {
          ...state.bookmarkedArticles,
          [payload.articleSlug]: payload.isBookmarked
        },
        isSubmitting: false
      };
    case REQUEST_SUBMITTED:
      return {
        ...state,
        isSubmitting: true
      };
    case REQUEST_FAILED:
      return {
        ...state,
        message: payload,
        isSubmitting: false
      };
    default:
      return state;
  }
};
