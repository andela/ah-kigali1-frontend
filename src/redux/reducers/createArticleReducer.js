import {
  NEW_ARTICLE,
  ARTICLE_ERROR,
  SUBMITTING_ARTICLE,
  FETCH_ARTICLE_TO_EDIT,
  ARTICLE_UPDATED,
  CLEAR_RESPONSE
} from "../actionTypes";

const initialState = {
  article: {
    title: "",
    body: "",
    description: "",
    tagsList: []
  }
};

const createArticleReducer = (state = initialState, action) => {
  switch (action.type) {
    case NEW_ARTICLE:
      return {
        ...state,
        article: { ...state.article, ...action.payload.article },
        message: action.payload.message,
        response: action.payload.article,
        isSubmitting: false
      };
    case ARTICLE_UPDATED:
      return {
        ...state,
        article: { ...state.article, ...action.payload.article },
        message: action.payload.message,
        response: action.payload.article,
        isSubmitting: false
      };

    case ARTICLE_ERROR:
      return {
        ...state,
        articleError: action.payload,
        isSubmitting: false
      };
    case SUBMITTING_ARTICLE:
      return {
        ...state,
        isSubmitting: true
      };
    case CLEAR_RESPONSE:
      return {
        ...state,
        message: null,
        articleError: null,
        response: null
      };
    case FETCH_ARTICLE_TO_EDIT:
      return {
        ...state,
        article: { ...state.article, ...action.payload.article }
      };
    default:
      return {
        ...state
      };
  }
};

export default createArticleReducer;
