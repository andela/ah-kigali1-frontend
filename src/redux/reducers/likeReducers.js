import {
  LIKE_ARTICLE,
  DISLIKE_ARTICLE,
  REQUEST_SUBMITTED,
  REQUEST_FAILED
} from "../actionTypes";

export const initialState = {
  likeCount: 0,
  isSubmitting: false,
  message: ""
};

export default (state = initialState, action) => {
  switch (action.type) {
    case LIKE_ARTICLE:
      return { ...state, isSubmitting: false, likeCount: state.likeCount + 1 };
    case DISLIKE_ARTICLE:
      return { ...state, isSubmitting: false, likeCount: state.likeCount - 1 };
    case REQUEST_SUBMITTED:
      return { ...state, isSubmitting: true };
    case REQUEST_FAILED:
      return { ...state, message: action.payload, isSubmitting: false };
    default:
      return state;
  }
};
