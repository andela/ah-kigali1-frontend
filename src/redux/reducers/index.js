import { combineReducers } from "redux";
import loginReducers from "./authReducers";
import resetPasswordReducers from "./resetPasswordReducers";
import updatePasswordReducers from "./updatePasswordReducers";
import socialAuthReducers from "./socialAuthReducers";
import createArticleReducer from "./createArticleReducer";
import readArticleReducer from "./readArticleReducer";
import userReducer from "./userReducer";
import searchReducers from "./searchReducers";
import following from "./followingReducer";
import notifier from "./notificationReducer";
import { reportArticle } from "./reportReducer";
import registration from "./registrationReducers";

export default combineReducers({
  auth: loginReducers,
  article: createArticleReducer,
  resetPassword: resetPasswordReducers,
  updatePassword: updatePasswordReducers,
  socialAuth: socialAuthReducers,
  fetchedArticle: readArticleReducer,
  user: userReducer,
  search: searchReducers,
  following,
  notifier,
  report: reportArticle,
  registration
});
