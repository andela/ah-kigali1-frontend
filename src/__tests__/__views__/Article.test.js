import React from "react";
import Enzyme, { mount } from "enzyme";
import toJson from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import {
  Article,
  mapStateToProps,
  mapDispatchToProps
} from "../../views/Article";
import { initialState } from "../../redux/reducers/likeReducers";

Enzyme.configure({ adapter: new Adapter() });

const setup = () => {
  const props = {
    like: jest.fn(),
    dislike: jest.fn(),
    likeCount: 0,
    isLiked: false
  };
  const enzymeWrapper = mount(<Article {...props} />);

  return {
    props,
    enzymeWrapper
  };
};

describe("tes like component", () => {
  const { props } = setup();
  const likeComponent = mount(<Article {...props} />);
  let instance;
  beforeEach(() => {
    instance = likeComponent.instance();
    jest.spyOn(instance.props, "like");
    jest.spyOn(instance.props, "dislike");
    instance.forceUpdate();
  });
  afterEach(() => {
    jest.restoreAllMocks();
  });
  it("snapshot testing for the component", () => {
    expect(toJson(likeComponent)).toMatchSnapshot();
    expect(likeComponent).toBeTruthy();
  });

  it("should render the component", () => {
    expect(likeComponent.find("div").length).toBe(3);
    expect(likeComponent.find("p").length).toBe(1);
    expect(likeComponent.find("p").text()).toEqual(props.likeCount.toString());
  });

  describe("handle event on like dislike button", () => {
    // test button color change

    it("should handle click like event ", () => {
      const likeButton = likeComponent
        .find("div")
        .at(1)
        .at(0)
        .find("img");
      likeButton.simulate("click", {});
      expect(instance.props.like).toHaveBeenCalled();
    });

    it("should handle click dislike event ", () => {
      props.isLiked = true;
      const likedComponent = mount(<Article {...props} />);
      const dislikeButton = likedComponent
        .find("div")
        .at(2)
        .at(0)
        .find("img");
      dislikeButton.simulate("click", {});
      expect(instance.props.dislike).toHaveBeenCalled();
    });
  });
  describe("test map state to props and map dispatch to props", () => {
    it("should test map state to props", () => {
      expect(mapStateToProps({ like: initialState })).toEqual({
        likeCount: 0,
        isSubmitting: false,
        message: "",
        isLiked: false
      });
    });
    it("should test map dispatch to props for like", () => {
      const dispatch = jest.fn();
      mapDispatchToProps(dispatch).like();
      expect(dispatch.mock.calls[0][0]).toBeInstanceOf(Function);
    });

    it("should test map dispatch to props for dislike", () => {
      const dispatch = jest.fn();
      mapDispatchToProps(dispatch).dislike();
      expect(dispatch.mock.calls[0][0]).toBeInstanceOf(Function);
    });
  });
});
