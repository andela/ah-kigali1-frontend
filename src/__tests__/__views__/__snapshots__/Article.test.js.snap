// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tes like component snapshot testing for the component 1`] = `
<Article
  dislike={[MockFunction]}
  isLiked={false}
  like={[MockFunction]}
  likeCount={0}
>
  <div>
    <LikeComponent
      icon="test-file-stub"
      onClick={[MockFunction]}
    >
      <div>
        <img
          alt="logo"
          className="share-icon"
          onClick={[MockFunction]}
          onKeyUp={[Function]}
          src="test-file-stub"
        />
         
      </div>
    </LikeComponent>
    <LikeComponent
      icon="test-file-stub"
      onClick={[Function]}
    >
      <div>
        <img
          alt="logo"
          className="share-icon"
          onClick={[Function]}
          onKeyUp={[Function]}
          src="test-file-stub"
        />
         
      </div>
    </LikeComponent>
    <p>
      0
    </p>
  </div>
</Article>
`;
